modules/ln_glgui/glgui.scm

```diff
@@ -227,15 +227,108 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+;;  (glgui-timings-set! frame-period-max: 0.5 frame-period-min: 0.05) -- 20/sec down to 2/sec
+;;  (glgui-timings-set! frame-period-custom: (lambda (x) 0.01))  -- constant 100/sec
+;;
+;;  When concerned about a precise frame rate take a hint from gambit
+;;  manual wrt. `thread-sleep!` and wait until a point in time no
+;;  matter how long it took to process events and render the GUI:
+;;
+;; (glgui-timings-set!
+;;  frame-period-custom:
+;;  (let ((last-frame-time (current-time-seconds))
+;;        (frame-period 0.05))
+;;    (lambda (x) ;; ignoring `x`
+;;      (let ((next (+ last-frame-time frame-period)))
+;;        (set! last-frame-time next)
+;;        (seconds->time next)))))
+
+(define glgui-timings-set!)
+
+;; glgui-wakeup! - a thunk, which when called will immediately unblock
+;; the thread waiting in glgui-event.  Should be called if other
+;; threads notice the event loop should proceed.  Immediately resets
+;; frame-period to frame-period-min.
+(define glgui-wakeup!)
+
 ;; process an input event
 ;; 20100519: allow multiple guis
 ;; 20100804: support gui offset
-(define (glgui-event guis t x0 y0)
-  (if (and glgui:active app:width app:height)
-    (let ((gs (if (list? guis) guis (list guis))))
-      (if (fx= t EVENT_REDRAW) (apply glgui:render gs)
-        (apply glgui:inputloop (append (list t x0 y0) gs)))
-   )))
+
+(define glgui-event
+  (let ((frame-period-max-value 0.5) ;; How long to sleep at most in redraw.
+        (step 0.05) ;; delay increase
+        (consecutive-redraw-count 1)
+        (customized-moment #f)
+        (wait-mutex (make-mutex 'glgui-event))
+        (wait-cv (make-condition-variable 'glgui-event)))
+    (define (timings-set! #!key (frame-period-max #f) (frame-period-min #f) (frame-period-custom #f))
+      (define (legal? x) (and (number? x) (positive? x)))
+      (if (legal? frame-period-max) (set! frame-period-max-value frame-period-max))
+      (if (legal? frame-period-min) (set! step frame-period-min))
+      (if (or (not frame-period-custom) (procedure? frame-period-custom))
+          (set! customized-moment frame-period-custom)))
+    (define (wakeup!)
+      (condition-variable-signal! wait-cv))
+    (define (reset-wait!)
+      (set! consecutive-redraw-count 1))
+    (define (wait-for-time-or-signal!)
+      ;; wait for delay or signal from other thread
+      (if (let ((moment (if customized-moment
+                            (customized-moment consecutive-redraw-count)
+                            (min frame-period-max-value (* consecutive-redraw-count step)))))
+              (mutex-unlock! wait-mutex wait-cv moment))
+          (reset-wait!)
+          (set! consecutive-redraw-count (fx+ consecutive-redraw-count 1))))
+    (define (glgui-event guis t x0 y0)
+      (if (and glgui:active app:width app:height)
+          (let ((gs (if (list? guis) guis (list guis))))
+            (if (fx= t EVENT_REDRAW)
+                (when (mutex-lock! wait-mutex 0)
```

Comment:
`when` is not available in Gambit 4.7.9, it was introduced only in 4.9.3. See #333. I suggest we change this to an `if` as otherwise this is a blocker to merging.